de.metas.adempiere.libero.libero/src/main/java/org/eevolution/process/PP_Product_BOM_Check.java

```diff
@@ -18,56 +18,142 @@
  *****************************************************************************/
 package org.eevolution.process;
 
-import static org.adempiere.model.InterfaceWrapperHelper.load;
+import java.util.concurrent.atomic.AtomicInteger;
 
+import org.adempiere.ad.dao.IQueryBL;
+import org.adempiere.ad.dao.IQueryBuilder;
 import org.adempiere.ad.trx.api.ITrxManager;
 import org.adempiere.exceptions.AdempiereException;
 import org.adempiere.model.InterfaceWrapperHelper;
 import org.adempiere.util.Services;
+import org.adempiere.util.StringUtils;
 import org.compiere.model.I_M_Product;
 import org.eevolution.api.IProductBOMBL;
 import org.eevolution.api.IProductBOMDAO;
 import org.eevolution.model.I_PP_Product_BOM;
 import org.eevolution.model.I_PP_Product_BOMLine;
 
+import de.metas.process.IProcessPrecondition;
+import de.metas.process.IProcessPreconditionsContext;
 import de.metas.process.JavaProcess;
+import de.metas.process.Param;
+import de.metas.process.ProcessPreconditionsResolution;
 import de.metas.process.RunOutOfTrx;
+import lombok.NonNull;
 
 /**
  * Title: Check BOM Structure (free of cycles) Description: Tree cannot contain BOMs which are already referenced
  *
  * @author Tony Snook (tspc)
  * @author Teo Sarca, SC ARHIPAC SERVICE SRL
  */
-public class PP_Product_BOM_Check extends JavaProcess
+public class PP_Product_BOM_Check extends JavaProcess implements IProcessPrecondition
 {
 	private final transient IProductBOMBL productBOMBL = Services.get(IProductBOMBL.class);
 	private final transient IProductBOMDAO productBOMDAO = Services.get(IProductBOMDAO.class);
 	private final transient ITrxManager trxManager = Services.get(ITrxManager.class);
 
+	@Param(parameterName = I_M_Product.COLUMNNAME_M_Product_Category_ID, mandatory = false)
+	private int p_M_Product_Category_ID;
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(IProcessPreconditionsContext context)
+	{
+		if (!(I_M_Product.Table_Name.equals(context.getTableName()) || I_PP_Product_BOM.Table_Name.equals(context.getTableName())))
+		{
+			return ProcessPreconditionsResolution.reject();
+		}
+
+		if (context.getSelectionSize()>1)
+		{
+			return ProcessPreconditionsResolution.reject();
+		}
+
+		return ProcessPreconditionsResolution.accept();
+	}
+
+
 	@Override
 	@RunOutOfTrx
 	protected String doIt()
 	{
 		log.info("Check BOM Structure");
 
+		if (p_M_Product_Category_ID > 0)
+		{
+			final IQueryBuilder<I_M_Product> queryBuilder = Services.get(IQueryBL.class)
+					.createQueryBuilder(I_M_Product.class, getCtx(), get_TrxName())
+					.addEqualsFilter(I_M_Product.COLUMNNAME_IsBOM, true);
+
+			if (p_M_Product_Category_ID > 0)
```

Comment:
this IF is not needed. we already know the product category Id > 0

Reply:
The idea with product category Id > 0 is to run the process form the menu, using product category.
If we drop the process from the menu, then we can drop product category.

Reply:
Got it. Thx!